experiments/process_sandbox/library_runner.cc

```diff
@@ -239,184 +219,531 @@ namespace sandbox
+      // The `upcall` function has a wake but not a wait because it is using
+      // the `wait` in this function, we need to ensure that we don't unbalance
+      // the wakes and waits.
+      if (new_depth == upcall_depth)
+      {
+        shared->token.is_child_executing = false;
+        shared->token.parent.wake();
+      }
+    } while (new_depth == upcall_depth);
+  }
+
+  SNMALLOC_SLOW_PATH
+  void bootstrap()
+  {
+    void* addr = nullptr;
+    size_t length = 0;
+    // Find the correct environment variables.  Note that libc is not fully
+    // initialised when this is called and so we have to be very careful about
+    // the libc function that we call.  We use the `environ` variable directly,
+    // rather than `getenv`, which may allocate memory.
+    //
+    // The parent process provides the shared memory object in the file
+    // descriptor with the number given by `SharedMemRegion` and the location
+    // where it should be mapped in an environment variable.  The child has to
+    // map this as the first step in bootstrapping (before most of libc
+    // initialises itself) to get a working heap.
+    for (char** e = environ; *e != nullptr; e++)
+    {
+      char* ev = *e;
+      const char ev_name[] = "SANDBOX_LOCATION=";
+      const size_t name_length = sizeof(ev_name) - 1;
+      if (strncmp(ev_name, ev, name_length) == 0)
+      {
+        ev += name_length;
+        char* end;
+        addr = reinterpret_cast<void*>(strtoull(ev, &end, 16));
+        SANDBOX_INVARIANT(
+          end[0] == ':',
+          "Expected ':' separator in environment variable, got '{}'",
+          end[0]);
+        length = strtoull(end + 1, nullptr, 16);
+        break;
+      }
+    }
+    // Abort if we weren't able to find the correct lengths.
+    if ((addr == nullptr) || (length == 0))
     {
-      set(reinterpret_cast<uintptr_t>(p), (uint8_t)size_bits, 1);
-      return;
+      DefaultPal::error("Unable to find memory location");
     }
-    set(reinterpret_cast<uintptr_t>(p), size_bits);
-    // Set redirect slide
-    uintptr_t ss = (uintptr_t)((size_t)p + SUPERSLAB_SIZE);
-    for (size_t i = 0; i < size_bits - SUPERSLAB_BITS; i++)
+
+    // fprintf(stderr, "Child starting\n");
+    // printf(
+    //"Child trying to map fd %d at addr %p (0x%zx)\n", SharedMemRegion, addr,
+    // length);
+    void* ptr = mmap(
+      addr,
+      length,
+      PROT_READ | PROT_WRITE,
+      MAP_FIXED_NOREPLACE | MAP_SHARED | platform::detail::map_nocore,
+      SharedMemRegion,
+      0);
+
+    // printf("%p\n", ptr);
+    if (ptr == MAP_FAILED)
+    {
+      err(1, "Mapping shared heap failed");
+    }
+
+    shared = reinterpret_cast<SharedMemoryRegion*>(ptr);
+    // Splice the pagemap page inherited from the parent into the pagemap.
+    void* pagemap_chunk = GlobalPagemap::pagemap().page_for_address(
+      reinterpret_cast<uintptr_t>(ptr));
+    munmap(pagemap_chunk, 4096);
```

Comment:
Magic number 4096. Would be nice to know why you're separating this chunk of the shared memory region.